### PR TITLE
BEP-404: Clear Miner History when Switching Validators Set

### PR DESCRIPTION
**Pull request by buddh0**
<pre>
  BEP: 404
  Title: Clear Miner History when Switching Validators Set
  Status: Candidate
  Type: Standards
  Created: 2024-07-01
</pre>

# BEP-404: Clear Miner History when Switching Validators Set

- [BEP-404: Clear Miner History when Switching Validators Set](#bep-404-clear-miner-history-when-switching-validators-set)
  - [1. Summary](#1-summary)
  - [2. Motivation](#2-motivation)
  - [3. Status](#3-status)
  - [4. Specification](#4-specification)
  - [5. Backwards Compatibility](#5-backwards-compatibility)
  - [6. License](#6-license)

## 1. Summary
After switching the validator set on the BSC chain, the previous block production record is still used to restrict some validators from producing blocks for the first few blocks after the switch, leading to brief network instability. This BEP aims to resolve this issue.

## 2. Motivation
The validator set on the BSC chain switches at each epoch.

Assume the number of validators in an epoch is validatorN. To ensure the consistency of the BSC chain, it is required that the chain must have at least validatorN/2+1 validators to continue advancing the chain.

To achieve this constraint, a set called MinerHistory is defined, with a length of validatorN/2, recording the validators that produced the last validatorN/2 blocks. New block-producing validators are required to not be in MinerHistory.

This BEP aims to enhance network stability by clearing MinerHistory when the validator set switches.

## 3. Status
This BEP is in progress.

## 4. Specification
In the previous specification, when the validator set switches at each epoch, MinerHistory is not cleared, causing some validators to be unable to produce blocks normally for the first few blocks after the switch.

This BEP clarifies that MinerHistory should be cleared when switching the validator set to solve this issue.

```Go
MinerHistory  map[uint64]Address  

// The conditions for switching remain unchanged
if BlockNum % EpochLength == (validatorN/2) {
  MinerHistory = Address{}
}
```

## 5. Backwards Compatibility
There are no known backward compatibility issues.

## 6. License
The content is licensed under [CC0](https://creativecommons.org/publicdomain/zero/1.0/)

### CODE REVIEW

**Review by unclezoro — DISMISSED** [review]

File: BEPs/BEP-404.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+In the previous specification, when the validator set switches at each epoch, MinerHistory is not cleared, causing some validators to be unable to produce blocks normally for the first few blocks after the switch.
+<div align="center">
+<img src=./assets/BEP-404/3-1.png width=100% />
+</div>
+This BEP clarifies that MinerHistory should be cleared when switching the validator set to solve this issue.
+
+```Go
+MinerHistory  map[uint64]Address  
+
+if switching validator set for new epoch {
```
Comment (zzzckck) [inline BEPs/BEP-404.md:46]: 1.pls be more specific when will this clear MinerHistory be executed?
Is it after processing a block with block number satisfy: 
```
BlockNum % EpochLength == (ValidatorSize/2)
```
2.Could you add a part about how to handle reorg?

File: README.md [hunk 2]
```diff
@@ -73,6 +73,7 @@ Here is the list of subjects of BEPs:
 | [BEP-362](./BEPs/BEP-362.md) | Greenfield Storage Fee Paymaster                          | Standards | Candidate |
 | [BEP-364](./BEPs/BEP-364.md) | Primary Storage Provider acts as the upload agent for object creation and update on Greenfield | Standards | Candidate |
 | [BEP-366](./BEPs/BEP-366.md) | PGreenfield Atomic Object Update                           | Standards | Candidate |
+| [BEP-403](./BEPs/BEP-403.md) | Clear Miner History when Switching Validator Set          | Standards | Candidate |
```
Comment (zzzckck) [inline README.md:76]: need to update

File: BEPs/assets/BEP-404/3-1.png
Comment (zzzckck) [inline BEPs/assets/BEP-404/3-1.png:1]: prefer Professional illustrations in BEP.

Comment (buddh0) [reply BEPs/assets/BEP-404/3-1.png:1]: > prefer Professional illustrations in BEP.

accepted

File: BEPs/BEP-404.md (hunk 1 above)
Comment (buddh0) [reply BEPs/BEP-404.md]: 1. first suggestion accepted
2. no change about the reorg logic

**Comment by zzzckck** [comment]
may refer this block as an example: https://bscscan.com/block/40110411

**Review by zzzckck — APPROVED** [review]

**Pull request merged**